File: Leetcode/41-First-Missing-Positive/FirstMissingPositive.go
func firstMissingPositive(nums []int) int {
	if nums == nil || len(nums) == 0 {
		return 1
	}
	size := len(nums)
	for i := 0; i < size; i++ {
		if nums[i] <= 0 {
			continue
		}
		if nums[i] > size {
			continue
		}
		if nums[nums[i]-1] == nums[i] {
			continue
		}
		temp := nums[i]
		nums[i], nums[temp-1] = nums[temp-1], nums[i]
		i--
	}

	for ind, val := range nums {
		if val != ind+1 {
			return ind + 1
		}
	}
	return size + 1
}

File: Lintcode/189-First-Missing-Positive/FirstMissingPositive.java
public class FirstMissingPositive {
	/**
	 * @param A
	 *            : an array of integers
	 * @return: an integer
	 */
	public int firstMissingPositive(int[] A) {
		if (A == null || A.length == 0) {
			return 1;
		}

		for (int i = 0; i < A.length; i++) {
			if (A[i] <= 0) {
				continue;
			}
			if (A[i] > A.length) {
				continue;
			}
			if (A[A[i] - 1] == A[i]) {
				continue;
			}
			int temp = A[i];
			A[i] = A[temp - 1];
			A[temp - 1] = temp;
			// pay attention to this! need to check A[i]`s new value in the next
			// iteration.
			i--;
		}

		for (int i = 0; i < A.length; i++) {
			if (A[i] != i + 1) {
				return i + 1;
			}
		}
		return A.length + 1;
	}
}
